cpp-ch/local-engine/Storages/Parquet/RowRanges.h

```diff
@@ -0,0 +1,208 @@
+            {
+                Range range1 = *it1_pair.first;
+                if (range1.isAfter(range2))
+                {
+                    result.add(range2);
+                    range2 = range1;
+                    std::swap(it1_pair, it2_pair);
+                }
+                else
+                {
+                    result.add(range1);
+                }
+                ++it1_pair.first;
+            }
+            result.add(range2);
+        }
+        else
+        {
+            it2_pair = it1_pair;
+        }
+        while (it2_pair.first != it2_pair.second)
+        {
+            result.add(*it2_pair.first);
+            ++it2_pair.first;
+        }
+
+        return result;
+    }
+
+    static RowRanges intersection(const RowRanges & left, const RowRanges & right)
+    {
+        RowRanges result;
+
+        int rightIndex = 0;
+        for (const Range & l : left.ranges)
+        {
+            for (int i = rightIndex, n = right.ranges.size(); i < n; ++i)
+            {
+                const Range & r = right.ranges[i];
+                if (l.isBefore(r))
+                {
+                    break;
+                }
+                else if (l.isAfter(r))
+                {
+                    rightIndex = i + 1;
+                    continue;
+                }
+                auto tmp = Range::intersection(l, r);
+                result.add(tmp.value());
+            }
+        }
+        return result;
+    }
+
+    RowRanges slice(const size_t from, const size_t to) const
+    {
+        RowRanges result;
+        for (const Range & range : ranges)
+            if (range.from >= from && range.to <= to)
+                result.add(range);
+        return result;
+    }
+
+    void add(const Range & range)
+    {
+        Range rangeToAdd = range;
+        for (int i = ranges.size() - 1; i >= 0; --i)
+        {
+            Range last = ranges[i];
+            assert(!last.isAfter(range));
+            const auto u = Range::unionRange(last, rangeToAdd);
+            if (!u.has_value())
+                break;
+            rangeToAdd = u.value();
+            ranges.erase(ranges.begin() + i);
+        }
+        ranges.push_back(rangeToAdd);
+    }
+
+    size_t rowCount() const
+    {
+        size_t cnt = 0;
+        for (const Range & range : ranges)
+            cnt += range.count();
+        return cnt;
+    }
+
+    bool isOverlapping(const size_t from, const size_t to) const
+    {
+        const Range searchRange(from, to);
+        const auto it = std::ranges::lower_bound(ranges, searchRange, [](const Range & r1, const Range & r2) { return r1.isBefore(r2); });
+        return it != ranges.end() && !it->isAfter(searchRange);
+    }
+
+    const std::vector<Range> & getRanges() const { return ranges; }
+
+    const Range & getRange(size_t index) const { return ranges[index]; }
+
+    std::string toString() const
+    {
+        std::string result = "[";
```

Comment:
should use `WriteBufferFromOwnString` ? Keep the taste same as `CH`

Reply:
This class eventually will be replaced in Arrow,  let's keep it